kmapper/kmapper.py

```diff
@@ -735,25 +742,38 @@ def visualize(
         if colorscale is None:
             colorscale = colorscale_default
 
-        # TODO:
-        #   - Make color functions more intuitive. How do they even work?
-        #   - Allow multiple color functions that can be toggled on and off.
+        if X_names is None:
+            X_names = []
+
+        if lens_names is None:
+            lens_names = []
 
         if not len(graph["nodes"]) > 0:
             raise Exception(
                 "Visualize requires a mapper with more than 0 nodes. \nIt is possible that the constructed mapper could have been constructed with bad parameters. This can occasionally happens when using the default clustering algorithm. Try changing `eps` or `min_samples` in the DBSCAN clustering algorithm."
             )
 
-        # Find the module absolute path and locate templates
-        module_root = os.path.join(os.path.dirname(__file__), "templates")
-        env = Environment(loader=FileSystemLoader(module_root))
-        color_values = init_color_values(graph, color_values)
+        if isinstance(color_function_name, str):
+            color_function_name = [color_function_name]
 
-        if X_names is None:
-            X_names = []
+        # test whether we have a color_function_names for each color_value vector
+        if color_values is not None and color_function_name is not None:
+            num_color_value_vectors = color_values.shape[1]
+            num_color_function_names = len(color_function_name)
+            if num_color_value_vectors != num_color_function_names:
+                raise Exception('{} `color_function_names` values found, but {} columns found in color_values. Must be equal.'.format(num_color_function_names, num_color_value_vectors))
 
-        if lens_names is None:
-            lens_names = []
+        if color_values is None:
+            if color_function_name is not None:
+                raise Exception('Parameter `color_function_name` was set, while `color_values` was not. Refusing to proceed.')
```

Comment:
Is it possible to proceed?  What about just having a `warn.warning` saying the colors might not be what you expect and then using the default generated `color_values` below?

Reply:
Hmm I can do a more nuanced warn / refusal test here. If color_values is None it'd be okay if color_function_name was set, as long as it was either a string or an array with a single entry. If that were the case, we could match that name up with the default-generated color_values. What can't happen on the html side is where the color_values has a different number of entries than the color_function_name -- I expect a one-to-one mapping when I generate the html, for switching functionality. 

But yeah, they could set color_function_name to override the 'Row number' label that would otherwise be set. I'll change that.

Reply:
This logic feels too verbose, but how about this general approach? https://github.com/scikit-tda/kepler-mapper/pull/196/commits/9e487b141b4fd20a30ebfde33582d12ffd0dfafa

Reply:
You're right it's not great. not sure if there is an easy solution as there are so many dependent parameters swimming around. 

Seems fine to leave it as is.

Reply:
Hmm I'll change `color_function_name` to be a default string argument. Immutable, so okay. Will cut down on logic.

Reply:
I tidied up the logic a bit, but I didn't do default string arg